File: front/src/types/skill.ts
export interface SkillListItem {
  slug: string;
  name: string;
  version: string;
  description?: string;
  category?: string;
  source_type: string;
  status: string;
  install_path: string;
  installed_at?: string;
  readme_available: boolean;
  bound_robot_count: number;
}

export interface SkillListResponse {
  total: number;
  items: SkillListItem[];
}

export interface SkillPromptFile {
  key: string;
  path: string;
  content: string;
}

export interface SkillBinding {
  robot_id: number;
  robot_name?: string;
  skill_slug: string;
  skill_name?: string;
  skill_version: string;
  category?: string;
  skill_description?: string;
  priority: number;
  status: string;
  prompt_keys: string[];
  binding_config: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface SkillDetail extends SkillListItem {
  manifest: Record<string, unknown>;
  readme_content?: string;
  prompts: SkillPromptFile[];
  bound_robots: SkillBinding[];
  installed_variants: SkillInstalledVariant[];
}

export interface SkillInstallResponse {
  message: string;
  skill: SkillDetail;
  install_task_id?: number;
}

export interface SkillBindingCreate {
  priority?: number;
  status?: string;
  binding_config?: Record<string, unknown>;
}

export interface SkillBindingUpdate {
  priority?: number;
  status?: string;
  binding_config?: Record<string, unknown>;
}

export interface SkillInstalledVariant {
  version: string;
  install_path: string;
  manifest_path: string;
  readme_available: boolean;
  is_current: boolean;
  installed_at?: string;
  prompt_keys: string[];
  bound_robot_count: number;
  bound_robot_ids: number[];
}

export interface SkillInstallTask {
  id: number;
  source_type: string;
  package_name?: string;
  package_url?: string;
  package_checksum?: string;
  package_signature?: string;
  signature_algorithm?: string;
  requested_by_user_id?: number;
  requested_by_username?: string;
  status: string;
  installed_skill_slug?: string;
  installed_skill_version?: string;
  error_message?: string;
  details: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  finished_at?: string;
}

export interface SkillInstallTaskListResponse {
  total: number;
  items: SkillInstallTask[];
}

export interface SkillInstallTaskActionResponse {
  message: string;
  task: SkillInstallTask;
}

export interface SkillAuditLog {
  id: number;
  action: string;
  target_type: string;
  status: string;
  actor_user_id?: number;
  actor_username?: string;
  actor_role?: string;
  robot_id?: number;
  skill_slug?: string;
  skill_version?: string;
  install_task_id?: number;
  message?: string;
  details: Record<string, unknown>;
  created_at: string;
}

export interface SkillAuditLogListResponse {
  total: number;
  items: SkillAuditLog[];
}

File: backend/app/schemas/skill.py
"""
Pydantic schemas for the skills bootstrap slice.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class SkillListItem(BaseModel):
    slug: str
    name: str
    version: str
    description: Optional[str] = None
    category: Optional[str] = None
    source_type: str = "local"
    status: str = "active"
    install_path: str
    installed_at: Optional[str] = None
    readme_available: bool = False
    bound_robot_count: int = 0


class SkillListResponse(BaseModel):
    total: int
    items: List[SkillListItem]


class SkillPromptFile(BaseModel):
    key: str
    path: str
    content: str


class SkillRobotBindingDetail(BaseModel):
    robot_id: int
    robot_name: Optional[str] = None
    skill_slug: str
    skill_name: Optional[str] = None
    skill_version: str
    category: Optional[str] = None
    skill_description: Optional[str] = None
    priority: int
    status: str
    prompt_keys: List[str] = Field(default_factory=list)
    binding_config: Dict[str, Any] = Field(default_factory=dict)
    provenance_install_task_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime


class SkillInstalledVariantInfo(BaseModel):
    version: str
    install_path: str
    manifest_path: str
    readme_available: bool = False
    is_current: bool = False
    installed_at: Optional[str] = None
    prompt_keys: List[str] = Field(default_factory=list)
    bound_robot_count: int = 0
    bound_robot_ids: List[int] = Field(default_factory=list)


class SkillDetail(SkillListItem):
    manifest: Dict[str, Any] = Field(default_factory=dict)
    readme_content: Optional[str] = None
    prompts: List[SkillPromptFile] = Field(default_factory=list)
    bound_robots: List[SkillRobotBindingDetail] = Field(default_factory=list)
    installed_variants: List[SkillInstalledVariantInfo] = Field(default_factory=list)


class SkillInstallResponse(BaseModel):
    message: str
    skill: SkillDetail
    install_task_id: Optional[int] = None


class SkillRemoteInstallRequest(BaseModel):
    package_url: str = Field(..., min_length=1)
    checksum: Optional[str] = None
    signature: Optional[str] = None
    signature_algorithm: Optional[str] = None


class SkillRemoteInstallResponse(BaseModel):
    message: str
    install_task_id: Optional[int] = None
    status: Optional[str] = None
    installed_skill_slug: Optional[str] = None
    installed_skill_version: Optional[str] = None


class SkillBindingCreate(BaseModel):
    priority: Optional[int] = Field(default=None, ge=1, le=9999)
    status: str = Field(default="active")
    binding_config: Dict[str, Any] = Field(default_factory=dict)
    install_task_id: Optional[int] = Field(default=None, ge=1)


class SkillBindingUpdate(BaseModel):
    priority: Optional[int] = Field(default=None, ge=1, le=9999)
    status: Optional[str] = None
    binding_config: Optional[Dict[str, Any]] = None
    install_task_id: Optional[int] = Field(default=None, ge=1)


class RuntimeSkillPromptBundle(BaseModel):
    active_skills: List[SkillRobotBindingDetail] = Field(default_factory=list)
    system_prompts: List[str] = Field(default_factory=list)
    retrieval_prompts: List[str] = Field(default_factory=list)
    answer_prompts: List[str] = Field(default_factory=list)


class SkillInstallTaskInfo(BaseModel):
    id: int
    source_type: str
    package_name: Optional[str] = None
    package_url: Optional[str] = None
    package_checksum: Optional[str] = None
    package_signature: Optional[str] = None
    signature_algorithm: Optional[str] = None
    requested_by_user_id: Optional[int] = None
    requested_by_username: Optional[str] = None
    status: str
    installed_skill_slug: Optional[str] = None
    installed_skill_version: Optional[str] = None
    error_message: Optional[str] = None
    details: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime
    finished_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class SkillInstallTaskListResponse(BaseModel):
    total: int
    items: List[SkillInstallTaskInfo]


class SkillInstallTaskActionResponse(BaseModel):
    message: str
    task: SkillInstallTaskInfo


class SkillAuditLogEntry(BaseModel):
    id: int
    action: str
    target_type: str
    status: str
    actor_user_id: Optional[int] = None
    actor_username: Optional[str] = None
    actor_role: Optional[str] = None
    robot_id: Optional[int] = None
    skill_slug: Optional[str] = None
    skill_version: Optional[str] = None
    install_task_id: Optional[int] = None
    message: Optional[str] = None
    details: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class SkillAuditLogListResponse(BaseModel):
    total: int
    items: List[SkillAuditLogEntry]
